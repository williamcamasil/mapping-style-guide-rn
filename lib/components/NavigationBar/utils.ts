import { ColorPaletesType } from '../../tokens';

export type NavigationBarVariantType =
  'dark-content' |
  'light-content';

export const getTextColor = (
  variant?: NavigationBarVariantType,
): ColorPaletesType => {
  switch (variant) {
    case 'light-content':
      return 'neutralWhite';
    case 'dark-content':
    default:
      return 'neutralGray700';
  }
};
