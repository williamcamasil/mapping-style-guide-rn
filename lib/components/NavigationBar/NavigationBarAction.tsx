import React, { ComponentType } from 'react';
import {
  StyleSheet, TouchableOpacity, TouchableOpacityProps,
} from 'react-native';

import { SvgProps } from 'react-native-svg';

import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';
import Spacer from '../Spacer';
import Text from '../Text';
import { NavigationBarVariantType } from './utils';

type NavigationBarActionPropTypes<IconProps extends SvgProps = SvgProps> = TouchableOpacityProps & {
  Icon: ComponentType<IconProps>;
  color?: ColorPaletesType;
  theme: AppThemeType;
  text?: string;
  variant?: NavigationBarVariantType;
};

const styles = StyleSheet.create({
  button: {
    flexDirection: 'row',
  },
});

const NavigationBarAction: React.FC<NavigationBarActionPropTypes> = ({
  Icon,
  theme,
  color,
  text,
  variant,
  ...others
}) => {
  const getDefaultActionColor = () => {
    if (variant === 'light-content') {
      return 'neutralWhite';
    }
    if (text) {
      return 'neutralGray600';
    }
    return 'neutralGray700';

  };

  const actionColor = color ?? getDefaultActionColor();

  const renderTextRight = () => {
    if (!text) return null;

    return (
      <>
        <Spacer size={theme.spacings.sXXS} />
        <Text
          testID="text-right"
          variant="body"
          color={actionColor}
        >
          {text}
        </Text>
      </>
    );
  };

  return (
    <TouchableOpacity activeOpacity={theme.opacities.intense} style={styles.button} {...others}>
      <Icon width={24} height={24} color={theme.colors[actionColor]} />
      {renderTextRight()}
    </TouchableOpacity>
  );
};

NavigationBarAction.defaultProps = {
  variant: 'dark-content',
};

export default withTheme(NavigationBarAction);
