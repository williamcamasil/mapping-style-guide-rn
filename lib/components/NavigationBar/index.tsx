import React, { useCallback } from 'react';
import {
  Platform,
  SafeAreaView,
  StatusBar, StatusBarProps, StyleSheet, View,
} from 'react-native';

import { useNavigationHolder } from 'mapping-context-rn';

import { Icons } from '../../assets';
import { useDimensions, useTextStyles, useViewStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import Divider from '../Divider';
import Spacer from '../Spacer';
import Text from '../Text';
import NavigationBarAction from './NavigationBarAction';
import { getTextColor, NavigationBarVariantType } from './utils';

const { OS } = Platform;

export type NavigationBarPropsType = {
  children?: React.ReactNode;
  onBackPress?: () => void;
  theme: AppThemeType;
  title?: string;
  rightContent?: React.ReactNode;
  closeIcon?: boolean;
  variant?: NavigationBarVariantType;
  statusBarTranslucent?: StatusBarProps['translucent'];
  statusBarBackgroundColor?: StatusBarProps['backgroundColor'];
  addDivider?: boolean;
};

const styles = StyleSheet.create({
  container: {
    height: 48,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  titleContainer: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    right: 0,
    left: 0,
    alignItems: 'center',
    justifyContent: 'center',
  },
  title: {
    textAlign: 'center',
  },
});

type NavigationBarType = React.FC<NavigationBarPropsType> & {
  Action: typeof NavigationBarAction;
};

const NavigationBar: NavigationBarType = ({
  onBackPress, theme, title, rightContent,
  children, closeIcon,
  variant = 'dark-content',
  statusBarTranslucent = true,
  statusBarBackgroundColor = 'transparent',
  addDivider = false,
}) => {
  const navigation = useNavigationHolder();
  const { statusBar } = useDimensions();

  const backButtonTouchableStyles = useTextStyles(() => [
    {
      marginLeft: theme.spacings.sLarge,
    },
  ], [theme]);

  const rightContentStyles = useTextStyles(() => [
    {
      marginRight: theme.spacings.sLarge,
    },
  ], [theme]);

  const childrenContainerStyles = useViewStyles(() => [
    {
      flex: 1,
      paddingHorizontal: theme.spacings.sLarge,
    },
  ], [theme]);

  const safeAreaStyles = useViewStyles(() => [
    {
      paddingTop: OS === 'android' ? statusBar.height : undefined,
    },
  ], [statusBar.height]);

  const handlePress = useCallback(() => {
    if (onBackPress) {
      onBackPress();
      return;
    }

    navigation.goBack();
  }, [navigation, onBackPress]);

  const renderContent = () => {
    if (children) {
      return (
        <View style={childrenContainerStyles}>
          {children}
        </View>
      );
    }

    return (
      <>
        <View style={styles.titleContainer} pointerEvents="none">
          <Text
            style={styles.title}
            variant="body"
            weight="bold"
            lineHeight="large"
            color={getTextColor(variant)}
            numberOfLines={2}
            testID="navigation-bar-title"
          >
            {title}
          </Text>
        </View>

        <View style={rightContentStyles}>
          {rightContent ? (
            rightContent
          ) : null}
        </View>
      </>
    );
  };

  const renderDivider = () => {
    if (!addDivider) return null;

    return (
      <>
        <Spacer size={theme.spacings.sXS} />
        <Divider color="neutralGray100" testID="divider-component" />
      </>
    );
  };

  return (
    <SafeAreaView style={safeAreaStyles}>
      <View style={styles.container}>
        <StatusBar
          barStyle={variant}
          translucent={statusBarTranslucent}
          backgroundColor={statusBarBackgroundColor}
        />
        <NavigationBarAction
          Icon={closeIcon ? Icons.Heavy.Close : Icons.Heavy.ArrowLeft}
          onPress={handlePress}
          style={backButtonTouchableStyles}
          testID="go-back-action"
          variant={variant}
        />
        {renderContent()}
      </View>
      {renderDivider()}
    </SafeAreaView>
  );
};

NavigationBar.Action = NavigationBarAction;

export default withTheme(NavigationBar);
