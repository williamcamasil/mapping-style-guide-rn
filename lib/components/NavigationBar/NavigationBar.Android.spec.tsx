import React from 'react';

import renderer from 'react-test-renderer';

import NavigationBar from '.';

beforeEach(() => {
  jest.clearAllMocks();
});

jest.mock('react-native/Libraries/Utilities/Platform', () => {
  const Platform = jest.requireActual(
    'react-native/Libraries/Utilities/Platform',
  );
  Platform.OS = 'android';
  return Platform;
});

describe('NavigationBar android snapshots', () => {
  it('should match snapshot default', () => {
    const tree = renderer
      .create(<NavigationBar
        title="Teste de título"
        statusBarTranslucent={false}
        statusBarBackgroundColor="#ff0000"
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with custom status bar', () => {
    const tree = renderer
      .create(<NavigationBar
        title="Teste de título"
        statusBarTranslucent={false}
        statusBarBackgroundColor="#ff0000"
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
