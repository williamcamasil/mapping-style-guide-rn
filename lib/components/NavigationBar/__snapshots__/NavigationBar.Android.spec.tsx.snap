// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBar android snapshots should match snapshot default 1`] = `
<View
  style={
    {
      "paddingTop": 42,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 48,
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "marginLeft": 28,
          "opacity": 1,
        }
      }
      testID="go-back-action"
    >
      <RNSVGSvgViewAndroid
        align="xMidYMid"
        bbHeight="24"
        bbWidth="24"
        fill="none"
        focusable={false}
        height={24}
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {
              "flex": 0,
              "height": 24,
              "width": 24,
            },
          ]
        }
        vbHeight={24}
        vbWidth={24}
        width={24}
      >
        <RNSVGGroup
          fill={null}
          propList={
            [
              "fill",
            ]
          }
        >
          <RNSVGPath
            clipRule={0}
            d="M12.778 3.222a1.1 1.1 0 0 1 0 1.556L5.556 12l7.222 7.222a1.1 1.1 0 1 1-1.555 1.556l-8-8a1.1 1.1 0 0 1 0-1.556l8-8a1.1 1.1 0 0 1 1.555 0Z"
            fill={
              {
                "payload": -12697271,
                "type": 0,
              }
            }
            fillRule={0}
            propList={
              [
                "fill",
                "fillRule",
              ]
            }
          />
        </RNSVGGroup>
      </RNSVGSvgViewAndroid>
    </View>
    <View
      pointerEvents="none"
      style={
        {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Text
        allowFontScaling={false}
        numberOfLines={2}
        style={
          {
            "color": "#3E4149",
            "flexShrink": 1,
            "fontFamily": "Averta CY",
            "fontSize": 16,
            "fontWeight": "bold",
            "letterSpacing": -0.02,
            "lineHeight": 22,
            "textAlign": "center",
          }
        }
        testID="navigation-bar-title"
      >
        Teste de título
      </Text>
    </View>
    <View
      style={
        {
          "marginRight": 28,
        }
      }
    />
  </View>
</View>
`;

exports[`NavigationBar android snapshots should match snapshot with custom status bar 1`] = `
<View
  style={
    {
      "paddingTop": 42,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 48,
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "marginLeft": 28,
          "opacity": 1,
        }
      }
      testID="go-back-action"
    >
      <RNSVGSvgViewAndroid
        align="xMidYMid"
        bbHeight="24"
        bbWidth="24"
        fill="none"
        focusable={false}
        height={24}
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {
              "flex": 0,
              "height": 24,
              "width": 24,
            },
          ]
        }
        vbHeight={24}
        vbWidth={24}
        width={24}
      >
        <RNSVGGroup
          fill={null}
          propList={
            [
              "fill",
            ]
          }
        >
          <RNSVGPath
            clipRule={0}
            d="M12.778 3.222a1.1 1.1 0 0 1 0 1.556L5.556 12l7.222 7.222a1.1 1.1 0 1 1-1.555 1.556l-8-8a1.1 1.1 0 0 1 0-1.556l8-8a1.1 1.1 0 0 1 1.555 0Z"
            fill={
              {
                "payload": -12697271,
                "type": 0,
              }
            }
            fillRule={0}
            propList={
              [
                "fill",
                "fillRule",
              ]
            }
          />
        </RNSVGGroup>
      </RNSVGSvgViewAndroid>
    </View>
    <View
      pointerEvents="none"
      style={
        {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Text
        allowFontScaling={false}
        numberOfLines={2}
        style={
          {
            "color": "#3E4149",
            "flexShrink": 1,
            "fontFamily": "Averta CY",
            "fontSize": 16,
            "fontWeight": "bold",
            "letterSpacing": -0.02,
            "lineHeight": 22,
            "textAlign": "center",
          }
        }
        testID="navigation-bar-title"
      >
        Teste de título
      </Text>
    </View>
    <View
      style={
        {
          "marginRight": 28,
        }
      }
    />
  </View>
</View>
`;
